File: FateGrandAutomata.Core/Modules/Game.cs
using CoreAutomata;

namespace FateGrandAutomata
{
    public static class Game
    {
        public const int ImageWidth = 1280,
            ImageHeight = 720,
            ScriptWidth = 2560,
            ScriptHeight = 1440;

        public static Region MenuScreenRegion { get; } = new Region(2100, 1200, 1000, 1000);
        public static Region ContinueRegion { get; } = new Region(1400, 1000, 600, 200);
        public static Region MenuStorySkipRegion { get; } = new Region(2240, 20, 300, 120);

        public static Location MenuSelectQuestClick { get; } = new Location(1900, 400);
        public static Location MenuStartQuestClick { get; } = new Location(2400, 1350);
        public static Location ContinueClick { get; } = new Location(1650, 1120);
        public static Location MenuStorySkipClick { get; } = new Location(2360, 80);
        public static Location MenuStorySkipYesClick { get; } = new Location(1600, 1100);

        // see docs/menu_boost_item_click_array.png
        public static Location MenuBoostItem1Click { get; } = new Location(1280, 418);
        public static Location MenuBoostItem2Click { get; } = new Location(1280, 726);
        public static Location MenuBoostItem3Click { get; } = new Location(1280, 1000);
        public static Location MenuBoostItemSkipClick { get; } = new Location(1652, 1304);

        public static Location[] MenuBoostItemClickArray { get;  } = new[]
        {
            MenuBoostItemSkipClick,
            MenuBoostItem1Click,
            MenuBoostItem2Click,
            MenuBoostItem3Click
        };

        public static Region StaminaScreenRegion { get; } = new Region(600, 200, 300, 300);
        public static Location StaminaOkClick { get; } = new Location(1650, 1120);
        public static Location StaminaSqClick { get; } = new Location(1270, 345);
        public static Location StaminaGoldClick { get; } = new Location(1270, 634);
        public static Location StaminaSilverClick { get; } = new Location(1270, 922);
        public static Location StaminaBronzeClick { get; } = new Location(1270, 1140);

        public static Region SupportScreenRegion { get; } = new Region(0, 0, 110, 332);
        public static Region SupportListRegion { get; } = new Region(70, 332, 378, 1091); // see docs/support_list_region.png
        public static Location SupportSwipeStartClick { get; } = new Location(35, 1190);
        public static Region SupportFriendsRegion { get; } = new Region(448, 332, 1210, 1091);

        public static Location SupportSwipeEndClick => Preferences.Instance.GameServer switch
        {
            GameServer.En => new Location(35, 390),
            GameServer.Jp => new Location(35, 350),
            GameServer.Cn => new Location(35, 390),
            GameServer.Tw => new Location(35, 390)
        };

        public static Region[] SupportListItemRegionArray { get;  } =
        {
            // see docs/support_list_item_regions_top.png
            new Region(76, 338, 2356, 428),
            new Region(76, 778, 2356, 390),

            // see docs/support_list_item_regions_bottom.png
            new Region(76, 558, 2356, 390),
            new Region(76, 991, 2356, 428)
        };

        public static Region SupportLimitBreakRegion { get; } = new Region(376, 0, 16, 90);
        public static Region SupportFriendRegion => new Region(2234, SupportListRegion.Y, 120, SupportListRegion.H); // see docs/friend_region.png

        public static Location SupportUpdateClick { get; } = new Location(1670, 250);
        public static Location SupportUpdateYesClick { get; } = new Location(1480, 1110);
        public static Location SupportListTopClick { get; } = new Location(2480, 360);
        public static Location SupportFirstSupportClick { get; } = new Location(1900, 500);

        public static Region BattleScreenRegion { get; } = new Region(2105, 1259, 336, 116); // see docs/battle_region.png

        public static Region BattleStageCountRegion => Preferences.Instance.GameServer switch
        {
            GameServer.En => new Region(1722, 25, 46, 53),
            GameServer.Jp => new Region(1722, 25, 46, 53),
            GameServer.Cn => new Region(1722, 25, 46, 53),
            GameServer.Tw => new Region(1710, 25, 55, 60)
        };

        public static Location BattleExtrainfoWindowCloseClick { get; } = new Location(2550, 0);
        public static Location BattleAttackClick { get; } = new Location(2300, 1200);
        public static Location BattleSkipDeathAnimationClick { get; } = new Location(1700, 100); // see docs/skip_death_animation_click.png

        // see docs/target_regions.png
        public static Region[] BattleTargetRegionArray { get; } =
        {
            new Region(0, 0, 485, 220),
            new Region(485, 0, 482, 220),
            new Region(967, 0, 476, 220)
        };

        public static Location[] BattleTargetClickArray { get; } =
        {
            new Location(90, 80),
            new Location(570, 80),
            new Location(1050, 80)
        };

        public static Location BattleSkill1Click { get; } = new Location(140, 1160);
        public static Location BattleSkill2Click { get; } = new Location(340, 1160);
        public static Location BattleSkill3Click { get; } = new Location(540, 1160);
        public static Location BattleSkill4Click { get; } = new Location(770, 1160);
        public static Location BattleSkill5Click { get; } = new Location(970, 1160);
        public static Location BattleSkill6Click { get; } = new Location(1140, 1160);
        public static Location BattleSkill7Click { get; } = new Location(1400, 1160);
        public static Location BattleSkill8Click { get; } = new Location(1600, 1160);
        public static Location BattleSkill9Click { get; } = new Location(1800, 1160);
        public static Location BattleSkillOkClick { get; } = new Location(1680, 850);

        public static Location BattleServant1Click { get; } = new Location(700, 880);
        public static Location BattleServant2Click { get; } = new Location(1280, 880);
        public static Location BattleServant3Click { get; } = new Location(1940, 880);

        public static Location BattleMasterSkillOpenClick { get; } = new Location(2380, 640);
        public static Location BattleMasterSkill1Click { get; } = new Location(1820, 620);
        public static Location BattleMasterSkill2Click { get; } = new Location(2000, 620);
        public static Location BattleMasterSkill3Click { get; } = new Location(2160, 620);

        public static Location BattleStartingMember1Click { get; } = new Location(280, 700);
        public static Location BattleStartingMember2Click { get; } = new Location(680, 700);
        public static Location BattleStartingMember3Click { get; } = new Location(1080, 700);
        public static Location BattleSubMember1Click { get; } = new Location(1480, 700);
        public static Location BattleSubMember2Click { get; } = new Location(1880, 700);
        public static Location BattleSubMember3Click { get; } = new Location(2280, 700);
        public static Location BattleOrderChangeOkClick { get; } = new Location(1280, 1260);

        public static Region[] BattleCardAffinityRegionArray { get; } = {
            // see docs/card_affinity_regions.png
            new Region(295, 650, 250, 200),
            new Region(810, 650, 250, 200),
            new Region(1321, 650, 250, 200),
            new Region(1834, 650, 250, 200),
            new Region(2348, 650, 250, 200)
        };

        public static Region[] BattleCardTypeRegionArray { get; } = {
            // see docs/card_type_regions.png
            new Region(0, 1060, 512, 200),
            new Region(512, 1060, 512, 200),
            new Region(1024, 1060, 512, 200),
            new Region(1536, 1060, 512, 200),
            new Region(2048, 1060, 512, 200)
        };

        public static Location[] BattleCommandCardClickArray { get; } = {
            new Location(300, 1000),
            new Location(750, 1000),
            new Location(1300, 1000),
            new Location(1800, 1000),
            new Location(2350, 1000),
        };

        public static Location[] BattleNpCardClickArray { get; } = {
            new Location(1000, 220),
            new Location(1300, 400),
            new Location(1740, 400)
        };

        public static Region ResultScreenRegion { get; } = new Region(100, 300, 700, 200);
        public static Region ResultBondRegion { get; } = new Region(2000, 820, 120, 120);
        public static Region ResultCeRewardRegion { get; } = new Region(1050, 1216, 33, 28);
        public static Location ResultCeRewardCloseClick { get; } = new Location(80, 60);
        public static Region ResultFriendRequestRegion { get; } = new Region(660, 120, 140, 160);
        public static Location ResultFriendRequestRejectClick { get; } = new Location(600, 1200);
        public static Region ResultQuestRewardRegion { get; } = new Region(1630, 140, 370, 250);
        public static Location ResultNextClick { get; } = new Location(2200, 1350); // see docs/quest_result_next_click.png
    }
}

File: app/src/main/java/com/mathewsachin/fategrandautomata/scripts/modules/Game.kt
package com.mathewsachin.fategrandautomata.scripts.modules

import com.mathewsachin.fategrandautomata.core.*
import com.mathewsachin.fategrandautomata.scripts.ImageLocator
import com.mathewsachin.fategrandautomata.scripts.enums.GameServerEnum
import com.mathewsachin.fategrandautomata.scripts.prefs.Preferences
import kotlin.time.seconds

class Game {
    companion object {
        val ImageSize = Size(1280, 720)
        val ScriptSize = Size(2560, 1440)

        val MenuScreenRegion = Region(2100, 1200, 1000, 1000)
        val ContinueRegion = Region(1400, 1000, 600, 200)
        val MenuStorySkipRegion = Region(2240, 20, 300, 120)

        val MenuSelectQuestClick = Location(2290, 440)
        val MenuStartQuestClick = Location(2400, 1350)
        val ContinueClick = Location(1650, 1120)
        val MenuStorySkipClick = Location(2360, 80)
        val MenuStorySkipYesClick = Location(1600, 1100)

        val RetryRegion = Region(1300, 1000, 700, 300)
        val WithdrawRegion = Region(400, 540, 1800, 190)
        val WithdrawAcceptClick = Location(1765, 720)

        // see docs/menu_boost_item_click_array.png
        val MenuBoostItem1Click = Location(1280, 418)
        val MenuBoostItem2Click = Location(1280, 726)
        val MenuBoostItem3Click = Location(1280, 1000)
        val MenuBoostItemSkipClick = Location(1652, 1304)

        val MenuBoostItemClickArray = listOf(
            MenuBoostItemSkipClick,
            MenuBoostItem1Click,
            MenuBoostItem2Click,
            MenuBoostItem3Click
        )

        val InventoryFullRegion = Region(1060, 910, 438, 70)

        val StaminaScreenRegion = Region(600, 200, 300, 300)
        val StaminaOkClick = Location(1650, 1120)
        val StaminaSqClick = Location(1270, 345)
        val StaminaGoldClick = Location(1270, 634)
        val StaminaSilverClick = Location(1270, 922)
        val StaminaBronzeClick = Location(1270, 1140)

        val SupportScreenRegion = Region(0, 0, 200, 400)
        val SupportListRegion = Region(70, 332, 378, 1091) // see docs/support_list_region.png
        val SupportSwipeStartClick = Location(35, 1190)
        val SupportSwipeEndClick = Location(5, 660)
        val SupportFriendsRegion = Region(448, 332, 1210, 1091)

        val SupportListItemRegionArray = listOf(
            // see docs/support_list_item_regions_top.png
            Region(76, 338, 2356, 428),
            Region(76, 778, 2356, 390),
            // see docs/support_list_item_regions_bottom.png
            Region(76, 558, 2356, 390),
            Region(76, 991, 2356, 428)
        )

        val SupportLimitBreakRegion = Region(376, 0, 16, 90)
        val SupportFriendRegion = Region(
            2234,
            SupportListRegion.Y,
            120,
            SupportListRegion.Height
        ) // see docs/friend_region.png

        val SupportUpdateClick = Location(1670, 250)
        val SupportUpdateYesClick = Location(1480, 1110)
        val SupportListTopClick = Location(2480, 360)
        val SupportFirstSupportClick = Location(1900, 500)

        val BattleScreenRegion = Region(2105, 1259, 336, 116) // see docs/battle_region.png

        val BattleStageCountRegion
            get() = when (Preferences.GameServer) {
                GameServerEnum.En -> Region(1722, 25, 46, 53)
                GameServerEnum.Jp -> Region(1722, 25, 46, 53)
                GameServerEnum.Cn -> Region(1722, 25, 46, 53)
                GameServerEnum.Tw -> Region(1710, 25, 55, 60)
            }

        val BattleExtrainfoWindowCloseClick = Location(2550, 10)
        val BattleAttackClick = Location(2300, 1200)

        // see docs/target_regions.png
        val BattleTargetRegionArray = listOf(
            Region(0, 0, 485, 220),
            Region(485, 0, 482, 220),
            Region(967, 0, 476, 220)
        )

        val BattleTargetClickArray = listOf(
            Location(90, 80),
            Location(570, 80),
            Location(1050, 80)
        )

        val BattleSkill1Click = Location(140, 1160)
        val BattleSkill2Click = Location(340, 1160)
        val BattleSkill3Click = Location(540, 1160)
        val BattleSkill4Click = Location(770, 1160)
        val BattleSkill5Click = Location(970, 1160)
        val BattleSkill6Click = Location(1140, 1160)
        val BattleSkill7Click = Location(1400, 1160)
        val BattleSkill8Click = Location(1600, 1160)
        val BattleSkill9Click = Location(1800, 1160)
        val BattleSkillOkClick = Location(1680, 850)

        val BattleServant1Click = Location(700, 880)
        val BattleServant2Click = Location(1280, 880)
        val BattleServant3Click = Location(1940, 880)

        val BattleMasterSkillOpenClick = Location(2380, 640)
        val BattleMasterSkill1Click = Location(1820, 620)
        val BattleMasterSkill2Click = Location(2000, 620)
        val BattleMasterSkill3Click = Location(2160, 620)

        val BattleStartingMember1Click = Location(280, 700)
        val BattleStartingMember2Click = Location(680, 700)
        val BattleStartingMember3Click = Location(1080, 700)
        val BattleSubMember1Click = Location(1480, 700)
        val BattleSubMember2Click = Location(1880, 700)
        val BattleSubMember3Click = Location(2280, 700)
        val BattleOrderChangeOkClick = Location(1280, 1260)

        val BattleCardAffinityRegionArray = listOf(
            // see docs/card_affinity_regions.png
            Region(295, 650, 250, 200),
            Region(810, 650, 250, 200),
            Region(1321, 650, 250, 200),
            Region(1834, 650, 250, 200),
            Region(2348, 650, 250, 200)
        )

        val BattleCardTypeRegionArray = listOf(
            // see docs/card_type_regions.png
            Region(0, 1060, 512, 200),
            Region(512, 1060, 512, 200),
            Region(1024, 1060, 512, 200),
            Region(1536, 1060, 512, 200),
            Region(2048, 1060, 512, 200)
        )

        val BattleCommandCardClickArray = listOf(
            Location(300, 1000),
            Location(750, 1000),
            Location(1300, 1000),
            Location(1800, 1000),
            Location(2350, 1000)
        )

        val BattleNpCardClickArray = listOf(
            Location(1000, 220),
            Location(1300, 400),
            Location(1740, 400)
        )

        val BattleServantFaceRegionArray = listOf(
            Region(106, 800, 300, 200),
            Region(620, 800, 300, 200),
            Region(1130, 800, 300, 200),
            Region(1644, 800, 300, 200),
            Region(2160, 800, 300, 200),

            Region(678, 190, 300, 200),
            Region(1138, 190, 300, 200),
            Region(1606, 190, 300, 200)
        )

        val ResultScreenRegion = Region(100, 300, 700, 200)
        val ResultBondRegion = Region(2000, 750, 120, 190)
        val ResultMasterExpRegion = Region(1290, 360, 380, 90)
        val ResultMatRewardsRegion = Region(2090, 1300, 260, 110)
        val ResultMasterLvlUpRegion = Region(2000, 170, 230, 250)

        val ResultCeRewardRegion = Region(1050, 1216, 33, 28)
        val ResultCeRewardCloseClick = Location(80, 60)
        val ResultFriendRequestRegion = Region(660, 120, 140, 160)
        val ResultFriendRequestRejectClick = Location(600, 1200)
        val ResultQuestRewardRegion = Region(1630, 140, 370, 250)
        val ResultNextClick = Location(2200, 1350) // see docs/quest_result_next_click.png

        fun needsToRetry() = RetryRegion.exists(ImageLocator.Retry)

        fun retry() {
            RetryRegion.click()

            AutomataApi.wait(2.seconds)
        }

        val GudaFinalRewardsRegion = Region(1160, 1040, 228, 76)
    }
}